chunk_downloader.go

```diff
@@ -630,3 +681,28 @@ func copyChunkStream(body io.Reader, rows chan<- []*string) error {
 	}
 	return nil
 }
+
+// ResultBatch object encapsulates a function that retrieves a subset of rows in a result set.
+type ResultBatch struct {
+	Rec                *[]array.Record
+	idx                int
+	scd                *snowflakeChunkDownloader
+	funcDownloadHelper func(context.Context, *snowflakeChunkDownloader, int) error
+}
+
+// Fetch returns an array of records representing a chunk in the query
+func (rb *ResultBatch) Fetch() error {
```

Comment:
I'm curious as to the design here: shouldn't this return the actual array.Record itself (along with an error), instead of storing the the record as an accessible object within the ResultBatch struct?

Reply:
My initial rationale was to not have a double memory burden but since the object contains `[]*array.Record`, it won't hurt to have it twice so that the user can Fetch it, but also re-use it in the future if necessary (double benefit). Changing it now.

Reply:
Does the object need to contain the record? Can it just be a product of the Fetch() function and nothing else? A close analogy would be returning a row object after a query, but you'd need to query the rows again in order to receive the same rows again; the driver won't do that for you.

Reply:
In this case we do need the `Rec` on the object because we are leveraging the `downloadChunkHelper` which only returns an `error`. It's analogous to how rows are downloaded in chunks within the `downloadChunkHelper` and then are applied to the `scd.Chunks` object, without explicitly being returned. I don't think we can omit that without having a major refactor on how the download workflow happens unfortunately. However, we can make `Rec` non-public (`rec`) if you rather not have it exposed by the object, but we thought it'd be nice for a user to have a pointer to it within the object itself in case it is needed (it's a cheap extra functionality) - up to you though if you'd rather have it lowercase!

Reply:
Thanks for the quick turnaround. I think the rec object might best be left private so the users can't control it and have fetch act as a getter (as you've done). 

It looks overall very good to me - I'm going to go over these changes today in a design review and come back with final feedback.

Reply:
sounds good. I will change it to private later this evening along with the clearer test. Consider it done for your design review!